### PR TITLE
Feature/dope 198 create extension for more types

### PR DESCRIPTION
**Pull request by jansigi**

### CODE REVIEW

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt [hunk 1]
```diff
@@ -3,22 +3,26 @@ package ch.ergon.dope.resolvable.expression.unaliased.type
 import ch.ergon.dope.resolvable.expression.TypeExpression
 import ch.ergon.dope.validtype.ValidType
 
-class Parameter<T : ValidType>(val previousValue: TypeExpression<T>, private var parameterName: String?) : TypeExpression<T> {
+class ValuedParameter<T : ValidType>(val value: TypeExpression<T>, private val parameterName: String?) : TypeExpression<T> {
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt:6]: Why the renaming to ValuedParameter?

Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt:6]: Can't we make parameterName non-nullable, and pass the parameter-count as a default if no name is provided?

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Primitive.kt [hunk 2]
```diff
@@ -49,13 +49,15 @@ class Primitive<T : ValidType> : TypeExpression<T> {
     }
 }
 
+fun TypeExpression<StringType>.isBlank(): Boolean = toQueryString().drop(1).dropLast(1).isBlank()
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Primitive.kt:52]: What is this logic?

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt [hunk 3]
```diff
@@ -3,22 +3,26 @@ package ch.ergon.dope.resolvable.expression.unaliased.type
...
         null -> "\$${ParameterCounter.count}"
         else -> "\$$parameterName"
     }
 }
 
-fun <T : ValidType> TypeExpression<T>.asParameter(parameterName: String) = Parameter(this, parameterName)
+fun <T : ValidType> TypeExpression<T>.asParameter(parameterName: String) = ValuedParameter(this, parameterName)
 
-fun <T : ValidType> TypeExpression<T>.asParameter() = Parameter(this, null)
+fun String.asParameter(parameterName: String) = ValuedParameter(this.toStringType(), parameterName)
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt:15]: Tests are missing here

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/EqualsExpression.kt [hunk 4]
```diff
@@ -1,16 +1,48 @@
...
 import ch.ergon.dope.validtype.BooleanType
+import ch.ergon.dope.validtype.NumberType
+import ch.ergon.dope.validtype.StringType
 import ch.ergon.dope.validtype.ValidType
 
-class EqualsExpression<T : ValidType>(
-    val left: TypeExpression<T>,
-    val right: TypeExpression<T>,
+class EqualsExpression<out T : ValidType>(
+    val left: TypeExpression<out T>,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/EqualsExpression.kt:14]: Do we really want _out_ here? Now we can write stuff like:
`EqualsExpression("5".toStringType(), 5.toNumberType())` and loose the type safety again. 
You are basically writing here: We can equal any two Valid-Types

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/GreaterOrEqualThanExpression.kt [hunk 5]
```diff
@@ -1,16 +1,38 @@
...
 import ch.ergon.dope.validtype.BooleanType
 import ch.ergon.dope.validtype.ComparableType
+import ch.ergon.dope.validtype.NumberType
+import ch.ergon.dope.validtype.StringType
 
 class GreaterOrEqualThanExpression<T : ComparableType>(
-    left: TypeExpression<T>,
-    right: TypeExpression<T>,
+    left: TypeExpression<out T>,
+    right: TypeExpression<out T>,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/GreaterOrEqualThanExpression.kt:14]: same comment as above

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt [hunk 6]
```diff
@@ -5,12 +5,12 @@ import ch.ergon.dope.resolvable.expression.unaliased.type.toStringType
...
+class NowStrExpression(
+    private val fmt: TypeExpression<StringType>,
+) : TypeExpression<StringType>, FunctionOperator {
+    override fun toQueryString(): String = toFunctionQueryString("NOW_STR", fmt)
 }
 
-fun nowStr(fmt: TypeExpression<StringType>): NowStrExpression = NowStrExpression(fmt)
-
-fun nowStr(): NowStrExpression = nowStr("".toStringType())
+fun nowStr(fmt: TypeExpression<StringType> = "".toStringType()): NowStrExpression = NowStrExpression(fmt)
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt:14]: Maybe add a comment referring to ticket Dope-177

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RepeatExpression.kt [hunk 7]
```diff
@@ -4,15 +4,20 @@ import ch.ergon.dope.resolvable.expression.TypeExpression
 import ch.ergon.dope.resolvable.expression.unaliased.type.toNumberType
 import ch.ergon.dope.resolvable.expression.unaliased.type.toStringType
 import ch.ergon.dope.resolvable.operator.FunctionOperator
+import ch.ergon.dope.validtype.NumberType
 import ch.ergon.dope.validtype.StringType
 
 class RepeatExpression(
     private val inStr: TypeExpression<StringType>,
-    private val n: Int,
+    private val n: TypeExpression<NumberType>,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RepeatExpression.kt:12]: Can we have a better parameter name than `n`?

File: crystal-map-connector/src/test/kotlin/ch/ergon/dope/CrystalMapAdapter.kt
Comment (pgruntz) [inline crystal-map-connector/src/test/kotlin/ch/ergon/dope/CrystalMapAdapter.kt:1]: I think there should be a lot more tests

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt (hunk 1 above)
Comment (sigristjan) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt:6]: Renamed to Parameter

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Primitive.kt (hunk 2 above)
Comment (sigristjan) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Primitive.kt:52]: A typeExpression always has a " ate the start and end of its queryString, so to check if it is blank i need to remove them

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt (hunk 3 above)
Comment (sigristjan) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt:15]: created ticket: DOPE-207

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt [hunk 8]
```diff
@@ -4,34 +4,43 @@ import ch.ergon.dope.resolvable.expression.TypeExpression
...
-            toFunctionQueryString(symbol = "LPAD", inStr, size.toNumberType(), extra)
-        }
+    override fun toQueryString(): String = toFunctionQueryString(symbol = "LPAD", inStr, size, extra = extra)
 }
 
-fun lpad(inStr: TypeExpression<StringType>, size: Int, extra: TypeExpression<StringType> = "".toStringType()) =
+fun lpad(
+    inStr: TypeExpression<StringType>,
+    size: TypeExpression<NumberType>,
+    extra: TypeExpression<StringType> = " ".toStringType(),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt:21]: what do you mean with `extra`? And why is there a space as a default value?

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt [hunk 9]
```diff
@@ -5,12 +5,12 @@ import ch.ergon.dope.resolvable.expression.unaliased.type.toStringType
 import ch.ergon.dope.resolvable.operator.FunctionOperator
 import ch.ergon.dope.validtype.StringType
 
-class NowStrExpression(private val fmt: TypeExpression<StringType>) : TypeExpression<StringType>, FunctionOperator {
-    override fun toQueryString(): String = this.toFunctionQueryString("NOW_STR", fmt)
+class NowStrExpression(
+    private val fmt: TypeExpression<StringType>,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt:9]: fmt?

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/EqualsExpression.kt (hunk 4 above)
Comment (sigristjan) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/EqualsExpression.kt:14]: No we dont want that here. Removed out

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt (hunk 6 above)
Comment (sigristjan) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt:14]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RepeatExpression.kt (hunk 7 above)
Comment (sigristjan) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/RepeatExpression.kt:12]: renamed to repeatAmount

File: crystal-map-connector/src/test/kotlin/ch/ergon/dope/CrystalMapAdapter.kt
Comment (sigristjan) [reply crystal-map-connector/src/test/kotlin/ch/ergon/dope/CrystalMapAdapter.kt:1]: created tocket: DOPE-207

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt (hunk 8 above)
Comment (sigristjan) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt:21]: https://docs.couchbase.com/cloud/n1ql/n1ql-language-reference/stringfun.html#fn-str-lpad
LPad has a thir parameter which is optional. The Third Param (extra) in Couchbase gets added to the left of the inStr if the size is bigger than the inStr length. And the extra is per default a space.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt (hunk 9 above)
Comment (sigristjan) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/NowStrExpression.kt:9]: stands for "format". renamed to format

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt (hunk 8 above)
Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt:21]: Can you rename 'extra' to something more meaningful (what it's actually used for) and maybe make it nullable if its optional

Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/stringfunction/LpadExpression.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt [hunk 10]
```diff
@@ -3,22 +3,26 @@ package ch.ergon.dope.resolvable.expression.unaliased.type
...
 import ch.ergon.dope.validtype.ValidType
 
-class Parameter<T : ValidType>(val previousValue: TypeExpression<T>, private var parameterName: String?) : TypeExpression<T> {
-    override fun toQueryString(): String = when (parameterName) {
-        null -> "\$${ParameterCounter.count}"
-        else -> "\$$parameterName"
-    }
+class Parameter<T : ValidType>(
+    val value: TypeExpression<T>,
+    private val parameterName: String = "${ParameterCounter.count}",
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt:8]: didn't we say we don't want to do that?

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/InExpression.kt [hunk 11]
```diff
@@ -15,3 +19,24 @@ class InExpression(
 
 fun <T : ValidType> TypeExpression<T>.inArray(array: TypeExpression<out ArrayType<out ValidType>>): InExpression =
     InExpression(this, array)
+
+fun Number.inArray(array: TypeExpression<out ArrayType<out ValidType>>): InExpression =
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/InExpression.kt:23]: I don't think we want this 'out'-statemets all over the place. It makes it less type safe. Here we want something like:
`fun Number.inArray(array: TypeExpression<ArrayType<NumberType>>): InExpression<NumberType> =`
Can you go over the whole project and remove these _out_ (maybe you have to add some more generic <T> on the way). I can only think of Arrays with different types where we might have to use an out.

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt (hunk 10 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/InExpression.kt (hunk 11 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/relational/InExpression.kt]: I deleted the ones i could fine and aren't necessary.

**Review by pgruntz — APPROVED** [review]

**Pull request merged**